usr/src/man/man1/ppgsz.1
'\" te
.\" Copyright (c) 2003, Sun Microsystems, Inc. All Rights Reserved
.\" The contents of this file are subject to the terms of the Common Development and Distribution License (the "License").  You may not use this file except in compliance with the License.
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE or http://www.opensolaris.org/os/licensing.  See the License for the specific language governing permissions and limitations under the License.
.\" When distributing Covered Code, include this CDDL HEADER in each file and include the License file at usr/src/OPENSOLARIS.LICENSE.  If applicable, add the following below this CDDL HEADER, with the fields enclosed by brackets "[]" replaced with your own identifying information: Portions Copyright [yyyy] [name of copyright owner]
.TH PPGSZ 1 "Jan 23, 2003"
.SH NAME
ppgsz \- set preferred page size for stack, heap, and/or other anonymous
segments
.SH SYNOPSIS
.LP
.nf
\fB/usr/bin/ppgsz\fR [\fB-F\fR] \fB-o\fR \fIoption\fR[,\fIoption\fR] \fIcmd\fR | \fB-p\fR \fIpid\fR...
.fi

.SH DESCRIPTION
.sp
.LP
The \fBppgsz\fR utility sets the preferred page size for stack, heap, and/or
other anonymous segments for the target process(es), that is, the launched
\fIcmd\fR or the process(es) in the pid list. \fBppgsz\fR stops the target
process(es) while changing the page size. See \fBmemcntl\fR(2).
.SH OPTIONS
.sp
.LP
The following options are supported:
.sp
.ne 2
.na
\fB\fB-F\fR\fR
.ad
.RS 22n
Force. Sets the preferred page size options(s) for target process(es) even if
controlled by other process(es). Caution should be exercised when using the
\fB-F\fR flag. See \fBproc\fR(1).
.RE

.sp
.ne 2
.na
\fB\fB-o\fR \fIoption\fR[,\fIoption\fR]\fR
.ad
.RS 22n
The \fIoption\fRs are:
.sp
.ne 2
.na
\fB\fBheap\fR=\fIsize\fR\fR
.ad
.RS 14n
This option specifies the preferred page size for the heap of the target
process(es). \fBheap\fR is defined to be the bss (uninitialized data) and the
brk area that immediately follows the bss (see \fBbrk\fR(2)). The preferred
heap page size is set for the existing heap and for any additional heap memory
allocated in the future. See NOTES.
.RE

.sp
.ne 2
.na
\fB\fBstack\fR=\fIsize\fR\fR
.ad
.RS 14n
This option specifies the preferred page size for the stack of the target
process(es). The preferred stack page size is set for the existing stack and
newly allocated parts of the stack as it expands.
.RE

.sp
.ne 2
.na
\fB\fBanon\fR=\fIsize\fR\fR
.ad
.RS 14n
This option specifies the preferred page size for all existing
\fBMAP_PRIVATE\fR anonymous segments of the target process(es),  other than
\fBheap\fR and \fBstack\fR, which are large enough to fit at least one aligned
page of the specified size. For the segments that are large enough, the
preferred page size is set starting at the first size-aligned address in the
segment. The \fBanon\fR preferred pagesize is not applied to \fBMAP_PRIVATE\fR
anonymous segments created in the future. See \fBMAP_ANON\fR in \fBmmap\fR(2).
.sp
Anonymous memory refers to \fBMAP_PRIVATE\fR pages that are not directly
associated with a file in some filesystem. The \fBppgsz\fR command uses
\fBmemcntl\fR(2) to set the preferred page size for anonymous segments. See
\fBMC_HAT_ADVISE\fR in \fBmemcntl\fR(2).
.RE

At least one of the above options must be specified.
.sp
\fIsize\fR must be a supported page size (see \fBpagesize\fR(1)) or \fB0\fR, in
which case the system will select an appropriate page size. See
\fBmemcntl\fR(2).
.sp
\fIsize\fR defaults to bytes and can be specified in octal (\fB0\fR), decimal,
or hexadecimal (\fB0x\fR). The numeric value can be qualified with \fBK\fR,
\fBM\fR, \fBG\fR, or \fBT\fR to specify Kilobytes, Megabytes, Gigabytes, or
Terabytes, respectively. \fB4194304\fR, \fB0x400000\fR, \fB4096K\fR,
\fB0x1000K\fR, and \fB4M\fR are different ways to specify 4 Megabytes.
.RE

.sp
.ne 2
.na
\fB\fB-p\fR \fIpid\fR\fR
.ad
.RS 22n
Sets the preferred page size option(s) for the target process(es) in the
process-id (\fIpid\fR) list following the \fB-p\fR option. The pid list can
also consist of names in the \fB/proc\fR directory. Only the process owner or
the super-user is permitted to set page size.
.sp
\fIcmd\fR is interpreted if \fB-p\fR is not specified. \fBppgsz\fR launches
\fIcmd\fR and applies page size option(s) to the new process.
.sp
The heap and stack preferred page sizes are inherited. Child process(es)
created (see \fBfork\fR(2)) from the launched process or the target process(es)
in the pid list after \fBppgsz\fR completes will inherit the preferred heap and
stack page sizes. The preferred page sizes of all segments are set back to the
default system page size on \fBexec\fR(2) (see \fBgetpagesize\fR(3C)). The
preferred page size for all other anonymous segments is not inherited by
children of the launched or target process(es).
.RE

.SH EXAMPLES
.LP
\fBExample 1 \fRSetting the preferred heap and stack page size
.sp
.LP
The following example sets the preferred heap page size to \fB4M\fR and the
preferred stack page size to \fB512K\fR for all \fBora\fR\(emowned processes
running commands that begin with \fBora\fR:

.sp
.in +2
.nf
example% \fBppgsz -o heap=4M,stack=512K -p `pgrep -u ora '^ora'`\fR
.fi
.in -2
.sp

.LP
\fBExample 2 \fRSetting the preferred anonymous page size
.sp
.LP
The following example sets the preferred page size of existing qualifying
anonymous segments to \fB512k\fR for process ID \fB953\fR:

.sp
.in +2
.nf
example% \fBppgsz -o anon=512k -p 953\fR
.fi
.in -2
.sp

.SH EXIT STATUS
.sp
.LP
If \fIcmd\fR is specified and successfully invoked (see \fBexec\fR(2)), the
exit status of \fBppgsz\fR will be the exit status of \fIcmd\fR. Otherwise,
\fBppgsz\fR will exit with one of the following values:
.sp
.ne 2
.na
\fB\fB0\fR \fR
.ad
.RS 7n
Successfully set preferred page size(s) for processes in the pid list.
.RE

.sp
.ne 2
.na
\fB\fB125\fR\fR
.ad
.RS 7n
An error occurred in \fBppgsz\fR. Errors include: invalid argument, invalid
page size(s) specified, and failure to set preferred page size(s) for one or
more processes in the pid list or \fIcmd\fR.
.RE

.sp
.ne 2
.na
\fB\fB126\fR\fR
.ad
.RS 7n
\fIcmd\fR was found but could not be invoked.
.RE

.sp
.ne 2
.na
\fB\fB127\fR\fR
.ad
.RS 7n
\fIcmd\fR could not be found.
.RE

.SH FILES
.sp
.ne 2
.na
\fB\fB/proc/*\fR \fR
.ad
.RS 29n
Process files.
.RE

.sp
.ne 2
.na
\fB\fB/usr/lib/ld/map.bssalign\fR \fR
.ad
.RS 29n
A template link-editor \fBmapfile\fR for aligning bss (see NOTES).
.RE

.SH ATTRIBUTES
.sp
.LP
See \fBattributes\fR(7) for descriptions of the following attributes:
.sp

.sp
.TS
box;
c | c
l | l .
ATTRIBUTE TYPE	ATTRIBUTE VALUE
_
Interface Stability	Evolving
.TE

.SH SEE ALSO
.sp
.LP
.BR ld (1),
.BR mpss.so.1 (1),
.BR pagesize (1),
.BR pgrep (1),
.BR pmap (1),
.BR proc (1),
.BR brk (2),
.BR exec (2),
.BR fork (2),
.BR memcntl (2),
.BR mmap (2),
.BR sbrk (2),
.BR getpagesize (3C),
.BR proc (5),
.BR attributes (7)
.sp
.LP
\fILinker and Libraries Guide\fR
.SH NOTES
.sp
.LP
Due to resource constraints, the setting of the preferred page size does not
necessarily guarantee that the target process(es) will get the preferred page
size. Use \fBpmap\fR(1) to view the \fBactual\fR heap and stack page sizes of
the target process(es) (see \fBpmap\fR \fB-s\fR option).
.sp
.LP
Large pages are required to be mapped at addresses that are multiples of the
size of the large page. Given that the heap is typically not large page
aligned, the starting portions of the heap (below the first large page aligned
address) are mapped with the system memory page size. See
\fBgetpagesize\fR(3C).
.sp
.LP
To provide a heap that will be mapped with a large page size, an application
can be built using a link-editor (\fBld\fR(1)) \fBmapfile\fR containing the
\fBbss\fR segment declaration directive. Refer to the section Mapfile Option in
the \fILinker and Libraries Guide\fR for more details of this directive and the
template \fBmapfile\fR provided in \fB/usr/lib/ld/map.bssalign\fR. Users are
cautioned that an alignment specification may be machine-specific and may lose
its benefit on different hardware platforms. A more flexible means of
requesting the most optimal underlying page size may evolve in future releases.
.sp
.LP
\fBmpss.so.1\fR(1), a preloadable shared object, can also be used to set the
preferred stack and/or heap page sizes.


usr/src/cmd/ppgsz/ppgsz.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License, Version 1.0 only
 * (the "License").  You may not use this file except in compliance
 * with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2001-2003 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 */

#pragma ident	"%Z%%M%	%I%	%E% SMI"

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <ctype.h>
#include <string.h>
#include <signal.h>
#include <errno.h>
#include <dirent.h>
#include <limits.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/mman.h>
#include <sys/wait.h>
#include <libproc.h>
#include <sys/sysmacros.h>
#include <libgen.h>
#include <thread.h>

#ifndef TRUE
#define	TRUE	1
#endif
#ifndef FALSE
#define	FALSE	0
#endif

static struct	ps_prochandle *Pr;
static char	*command;
static volatile int interrupt;
static int	Fflag;
static int	cflag = 1;

static void	intr(int);
static int	setpgsz(struct ps_prochandle *, int, size_t *);
static int	setpgsz_anon(struct ps_prochandle *, size_t, int);
static caddr_t	setup_mha(uint_t, size_t, int);
static size_t	discover_optimal_pagesize(struct ps_prochandle *,
		uint_t, pid_t);
static void	usage();

#define	INVPGSZ		3

/* subopt */

static char	*suboptstr[] = {
	"heap",
	"stack",
	"anon",
	NULL
};

enum	suboptenum {
	E_HEAP,
	E_STACK,
	E_ANON
};

static size_t
atosz(char *optarg)
{
	size_t		sz = 0;
	char		*endptr;

	if (optarg == NULL || optarg[0] == '\0')
		return (INVPGSZ);

	sz = strtoll(optarg, &endptr, 0);

	switch (*endptr) {
	case 'T':
	case 't':
		sz *= 1024;
	/*FALLTHRU*/
	case 'G':
	case 'g':
		sz *= 1024;
	/*FALLTHRU*/
	case 'M':
	case 'm':
		sz *= 1024;
	/*FALLTHRU*/
	case 'K':
	case 'k':
		sz *= 1024;
	/*FALLTHRU*/
	case 'B':
	case 'b':
	default:
		break;
	}
	return (sz);
}

/* pgsz array sufficient for max page sizes */

static size_t	pgsza[8 * sizeof (void *)];
static int	nelem;

static void
getpgsz()
{
	if ((nelem = getpagesizes(NULL, 0)) == 0) {
		(void) fprintf(stderr, "%s: cannot determine system page"
		    " sizes\n", command);
		exit(125);
	}

	(void) getpagesizes(pgsza, nelem);
}

static size_t
cnvpgsz(char *optarg)
{
	size_t		pgsz = atosz(optarg);
	int		i;

	if (!ISP2(pgsz) || ((pgsz < pgsza[0]) && pgsz != 0)) {
		pgsz = INVPGSZ;
	} else {
		for (i = nelem - 1; i >= 0; i--) {
			if (pgsz == pgsza[i])
				break;
			if (pgsz > pgsza[i]) {
				pgsz = INVPGSZ;
				break;
			}
		}
	}
	if (pgsz == INVPGSZ) {
		if (optarg != NULL) {
			(void) fprintf(stderr,
			    "%s: invalid page size specified (%s)\n",
			    command, optarg);
		} else {
			usage();
		}
		exit(125);
	}
	return (pgsz);
}

static void
usage()
{
	(void) fprintf(stderr,
	    "usage:\t%s -o option[,option] [-F] cmd | -p pid ...\n"
	    "    (set preferred page size of cmd or each process)\n"
	    "    -o option[,option]: options are\n"
	    "         stack=sz\n"
	    "         heap=sz\n"
	    "         anon=sz		(sz: valid page size or 0 (zero))\n"
	    "    -F: force grabbing of the target process(es)\n"
	    "    cmd: launch command\n"
	    "    -p pid ...: process id list\n",
	    command);
	exit(125);
}

int
main(int argc, char *argv[])
{
	int		rc, err = 0;
	int		opt, subopt;
	int		errflg = 0;
	char		*options, *value;
	size_t		pgsz[] = {INVPGSZ, INVPGSZ, INVPGSZ};
	pid_t		pid;
	int		status;

	if ((command = strrchr(argv[0], '/')) != NULL)
		command++;
	else
		command = argv[0];

	getpgsz();

	/* options */
	while ((opt = getopt(argc, argv, "o:Fp")) != EOF) {
		switch (opt) {
		case 'o':		/* options */
			options = optarg;
			while (*options != '\0') {
				subopt = getsubopt(&options, suboptstr, &value);
				switch (subopt) {
				case E_HEAP:
				case E_STACK:
				case E_ANON:
					pgsz[subopt] = cnvpgsz(value);
					break;
				default:
					errflg = 1;
					break;
				}
			}
			break;
		case 'F':		/* force grabbing (no O_EXCL) */
			Fflag = PGRAB_FORCE;
			break;
		case 'p':
			cflag = 0;
			break;
		default:
			errflg = 1;
			break;
		}
	}

	argc -= optind;
	argv += optind;

	if ((pgsz[E_HEAP] == INVPGSZ && pgsz[E_STACK] == INVPGSZ &&
	    pgsz[E_ANON] == INVPGSZ) || errflg || argc <= 0) {
		usage();
	}

	/* catch signals from terminal */
	if (sigset(SIGHUP, SIG_IGN) == SIG_DFL)
		(void) sigset(SIGHUP, intr);
	if (sigset(SIGINT, SIG_IGN) == SIG_DFL)
		(void) sigset(SIGINT, intr);
	if (sigset(SIGQUIT, SIG_IGN) == SIG_DFL)
		(void) sigset(SIGQUIT, intr);
	(void) sigset(SIGTERM, intr);

	if (cflag && !interrupt) {		/* command */
		int		err;
		char		path[PATH_MAX];

		Pr = Pcreate(argv[0], &argv[0], &err, path, sizeof (path));
		if (Pr == NULL) {
			switch (err) {
			case C_PERM:
				(void) fprintf(stderr,
				    "%s: cannot control set-id or "
				    "unreadable object file: %s\n",
				    command, path);
				break;
			case C_LP64:
				(void) fprintf(stderr,
				    "%s: cannot control _LP64 "
				    "program: %s\n", command, path);
				break;
			case C_NOEXEC:
				(void) fprintf(stderr, "%s: cannot execute "
				    "program: %s\n", command, argv[0]);
				exit(126);
				break;
			case C_NOENT:
				(void) fprintf(stderr, "%s: cannot find "
				    "program: %s\n", command, argv[0]);
				exit(127);
				break;
			case C_STRANGE:
				break;
			default:
				(void) fprintf(stderr,
				    "%s: %s\n", command, Pcreate_error(err));
				break;
			}
			exit(125);
		}

		if ((rc = setpgsz(Pr, Pstatus(Pr)->pr_dmodel, pgsz)) != 0) {
			(void) fprintf(stderr, "%s: set page size "
			    "failed for program: %s\n", command, argv[0]);
			(void) pr_exit(Pr, 1);
			exit(125);
		}

		/*
		 * release the command to run, wait for it and
		 * return it's exit status if we can.
		 */
		Prelease(Pr, 0);
		do {
			pid = wait(&status);
		} while (pid == -1 && errno == EINTR);

		if (pid == -1) {
			(void) fprintf(stderr, "%s: wait() error: %s\n",
			    command, strerror(errno));
			exit(125);
		}

		/*
		 * Pass thru the child's exit value.
		 */
		if (WIFEXITED(status))
			exit(WEXITSTATUS(status));
		exit(status | WCOREFLG);
	}

	/* process pids */

	while (--argc >= 0 && !interrupt) {
		char *arg;
		psinfo_t psinfo;
		int gret;

		(void) fflush(stdout);	/* line-at-a-time */

		/* get the specified pid and the psinfo struct */
		arg = *argv++;
		pid = proc_arg_psinfo(arg, PR_ARG_PIDS, &psinfo, &gret);

		if (pid == -1) {
			(void) fprintf(stderr, "%s: cannot examine pid %s:"
			    " %s\n", command, arg, Pgrab_error(gret));
			if (!isdigit(arg[0]) && strncmp(arg, "/proc/", 6)) {
				(void) fprintf(stderr,
				    "\tdo not use -p option"
				    " to launch a command\n");
			}
			err++;
		} else if ((Pr = Pgrab(pid, Fflag, &gret)) != NULL) {
			rc = setpgsz(Pr, Pstatus(Pr)->pr_dmodel, pgsz);
			if (rc != 0) {
				(void) fprintf(stderr, "%s: set page size "
				    "failed for pid: %d\n", command, (int)pid);
				err++;
			}
			Prelease(Pr, 0);
			Pr = NULL;
		} else {
			switch (gret) {
			case G_SYS:
				proc_unctrl_psinfo(&psinfo);
				(void) fprintf(stderr, "%s: cannot set page "
				    "size for system process: %d [ %s ]\n",
				    command, (int)pid, psinfo.pr_psargs);
				err++;
				break;
			case G_SELF:
				/* do it to own self */
				rc = setpgsz(NULL, psinfo.pr_dmodel, pgsz);
				if (rc != 0) {
					(void) fprintf(stderr, "%s: set page"
					    "size failed for self: %d\n",
					    command, (int)pid);
					err++;
				}
				break;
			default:
				(void) fprintf(stderr, "%s: %s: %d\n",
				    command, Pgrab_error(gret), (int)pid);
				err++;
				break;
			}
		}
	}

	if (interrupt || err)
		exit(125);

	return (0);
}

/* ARGSUSED */
static void
intr(int sig)
{
	interrupt = 1;
}

/* ------ begin specific code ------ */

/* set process page size */
/*ARGSUSED*/
static int
setpgsz(struct	ps_prochandle *Pr, int dmodel, size_t pgsz[])
{
	int			rc;
	int			err = 0;
	caddr_t			mpss;
	int			i;
	static uint_t	pgszcmd[] =
	{MHA_MAPSIZE_BSSBRK, MHA_MAPSIZE_STACK, MHA_MAPSIZE_VA};

	for (i = E_HEAP; i <= E_ANON; i++) {
		if (pgsz[i] == INVPGSZ)
			continue;

		if (i == E_ANON)
			rc = setpgsz_anon(Pr, pgsz[i], dmodel);
		else {
			mpss = setup_mha(pgszcmd[i], pgsz[i], dmodel);
			rc = pr_memcntl(Pr, NULL, 0, MC_HAT_ADVISE, mpss, 0, 0);
		}

		if (rc < 0) {
			(void) fprintf(stderr, "%s: warning: set %s page size "
			    "failed (%s) for pid %d\n", command, suboptstr[i],
			    strerror(errno), (int)Pstatus(Pr)->pr_pid);
			err++;
		}
	}
	return (err);
}


/*
 * Walk through the process' address space segments.  Set all anonymous
 * segments to the new page size.
 */
static int
setpgsz_anon(struct ps_prochandle *Pr, size_t pgsz, int dmodel)
{
	caddr_t		mpss;
	prmap_t		map;
	uintptr_t	addr;
	size_t		size;
	const psinfo_t	*psinfo;
	const pstatus_t	*pstatus;
	int		fd;
	int		rc;
	char		path[PATH_MAX];

	/*
	 * Setting the page size for anonymous segments on a process before it
	 * has run will have no effect, since it has not configured anonymous
	 * memory and the page size setting is not "sticky" inside the kernel.
	 * Any anonymous memory subsequently mapped will have the default page
	 * size.
	 */
	if (cflag)
		return (0);

	if ((psinfo = Ppsinfo(Pr)) == NULL)
		return (-1);
	if ((pstatus = Pstatus(Pr)) == NULL)
		return (-1);

	if (pgsz == 0)
		pgsz = discover_optimal_pagesize(Pr, dmodel, psinfo->pr_pid);

	mpss = setup_mha(MHA_MAPSIZE_VA, pgsz, dmodel);

	(void) snprintf(path, PATH_MAX, "/proc/%d/map", (int)psinfo->pr_pid);
	if ((fd = open(path, O_RDONLY)) < 0)
		return (-1);

	while (read(fd, &map, sizeof (map)) == sizeof (map)) {
		if ((map.pr_mflags & MA_ANON) == 0) {
			/* Not anon. */
			continue;
		} else if (map.pr_mflags & MA_SHARED) {
			/* Can't change pagesize for shared mappings. */
			continue;
		} else if (map.pr_vaddr + map.pr_size >
		    pstatus->pr_brkbase &&
		    map.pr_vaddr <
		    pstatus->pr_brkbase + pstatus->pr_brksize) {
			/* Heap. */
			continue;
		} else if (map.pr_vaddr >= pstatus->pr_stkbase &&
		    map.pr_vaddr + map.pr_size <=
		    pstatus->pr_stkbase + pstatus->pr_stksize) {
			/* Stack. */
			continue;
		} else if (map.pr_size < pgsz) {
			/* Too small. */
			continue;
		}

		/*
		 * Find the first address in the segment that is page-aligned.
		 */
		if (pgsz == 0 || ((map.pr_vaddr % pgsz) == 0))
			addr = map.pr_vaddr;
		else
			addr = map.pr_vaddr + (pgsz - (map.pr_vaddr % pgsz));

		/*
		 * Calculate how many pages will fit in the segment.
		 */
		if (pgsz == 0)
			size = map.pr_size;
		else
			size = map.pr_size - (addr % map.pr_vaddr) -
			    ((map.pr_vaddr + map.pr_size) % pgsz);

		/*
		 * If no aligned pages fit in the segment, ignore it.
		 */
		if (size < pgsz) {
			continue;
		}

		rc = pr_memcntl(Pr, (caddr_t)addr, size,
		    MC_HAT_ADVISE, mpss, 0, 0);

		/*
		 * If an error occurs on any segment, report the error here and
		 * then go on to try setting the page size for the remaining
		 * segments.
		 */
		if (rc < 0) {
			(void) fprintf(stderr, "%s: warning: set page size "
			    "failed (%s) for pid %d for anon segment at "
			    "address: %p\n", command, strerror(errno),
			    (int)psinfo->pr_pid, (void *)map.pr_vaddr);
		}
	}

	(void) close(fd);
	return (0);
}

/*
 * Discover the optimal page size for the process.
 * Do this by creating a 4M segment in the target process, set its pagesize
 * to 0, and read the map file to discover the page size selected by the system.
 */
static size_t
discover_optimal_pagesize(struct ps_prochandle *Pr, uint_t dmodel, pid_t pid)
{
	size_t			size = 0;
	size_t			len = pgsza[nelem - 1];
	prxmap_t		xmap;
	caddr_t			mha;
	void			*addr;
	int			fd = -1;
	char			path[PATH_MAX];

	(void) snprintf(path, PATH_MAX, "/proc/%d/xmap", (int)pid);
	if ((fd = open(path, O_RDONLY)) < 0)
		return (size);

	if ((addr = pr_mmap(Pr, (void *)len, len, PROT_READ | PROT_WRITE,
	    MAP_PRIVATE | MAP_ANON | MAP_ALIGN, -1, 0)) == MAP_FAILED) {
		goto err;
	}

	mha = setup_mha(MHA_MAPSIZE_VA, 0, dmodel);
	if (pr_memcntl(Pr, addr, len, MC_HAT_ADVISE, mha, 0, 0) < 0) {
		goto err;
	}

	/*
	 * Touch a page in the segment so the hat mapping gets created.
	 */
	(void) Pwrite(Pr, &len, sizeof (len), (uintptr_t)addr);

	/*
	 * Read through the address map looking for our segment.
	 */

	while (read(fd, &xmap, sizeof (xmap)) == sizeof (xmap)) {
		if (xmap.pr_vaddr == (uintptr_t)addr)
			break;
	}
	if (xmap.pr_vaddr != (uintptr_t)addr)
		goto err;

	size = xmap.pr_hatpagesize;

err:
	if (addr != MAP_FAILED) {
		if (pr_munmap(Pr, addr, len) == -1) {
			(void) fprintf(stderr,
			    "%s: couldn't delete segment at %p\n",
			    command, addr);
		}
	}
	if (fd != -1)
		(void) close(fd);

	return (size);
}

static struct memcntl_mha	gmha;
#ifdef _LP64
static struct memcntl_mha32	gmha32;
#endif

static caddr_t
/* ARGSUSED */
setup_mha(uint_t command, size_t pagesize, int dmodel)
{
#ifdef _LP64
	if (dmodel == PR_MODEL_ILP32) {
		gmha32.mha_cmd = command;
		gmha32.mha_flags = 0;
		gmha32.mha_pagesize = pagesize;
		return ((caddr_t)&gmha32);
	}
#endif
	gmha.mha_cmd = command;
	gmha.mha_flags = 0;
	gmha.mha_pagesize = pagesize;
	return ((caddr_t)&gmha);
}
